File: testdata/fixtures/kotlin/CWE-319.kt
// CWE-319: the address says http, so the request goes out in the clear.
//
// On a phone the network is somebody else's: a café access point, a captive
// portal, a carrier's transparent proxy. A cleartext request can be read and
// rewritten by all of them, and what comes back is whatever they sent - so
// this is not only a disclosure, it is a way to feed the app a response.
//
// The address is the whole of the evidence. A host that is only reachable on
// the loopback interface, or on the emulator's address for it, is not on
// anybody's network, and an XML namespace is not an address at all.
package fixture

import java.net.URL
import okhttp3.Request

class Api {

    fun fetch(): URL {
        // ruleid: kotlin-cleartext-http-url
        return URL("http://api.example.com/v1/orders")
    }

    fun request(): Request {
        // ruleid: kotlin-cleartext-http-url
        return Request.Builder().url("http://api.example.com/v1/orders").build()
    }

    fun secure(): URL {
        // ok: kotlin-cleartext-http-url
        return URL("https://api.example.com/v1/orders")
    }

    fun localDebug(): URL {
        // ok: kotlin-cleartext-http-url
        return URL("http://127.0.0.1:8080/health")
    }

    fun emulator(): URL {
        // ok: kotlin-cleartext-http-url
        return URL("http://10.0.2.2:8080/health")
    }

    fun namespace(): String {
        // ok: kotlin-cleartext-http-url
        return "http://schemas.android.com/apk/res/android"
    }
}

File: testdata/fixtures/swift/CWE-319.swift
// CWE-319: the request goes out in the clear.
//
// A phone is on somebody else's network almost all the time - a café access
// point, a captive portal, a carrier's transparent proxy - and every one of
// them can read a cleartext request. The half people forget is the other
// direction: whoever can read it can also rewrite the reply, so this is a way
// to feed the app whatever the network decides to say.
//
// App Transport Security blocks this by default, which means an app doing it
// has an `NSAllowsArbitraryLoads` in its Info.plist to match - two findings
// for one decision, and both worth having.
import Foundation

enum API {

    // ruleid: swift-cleartext-http-url
    static let base = URL(string: "http://api.example.com/v1")!

    static func fetch(_ session: URLSession) {
        // ruleid: swift-cleartext-http-url
        var request = URLRequest(url: URL(string: "http://api.example.com/login")!)
        request.httpMethod = "POST"
        session.dataTask(with: request).resume()
    }

    static func components() -> URLComponents? {
        // ruleid: swift-cleartext-http-url
        return URLComponents(string: "http://metrics.example.com/collect")
    }

    static func secure() {
        // ok: swift-cleartext-http-url
        _ = URL(string: "https://api.example.com/v1")
    }

    static func loopback() {
        // ok: swift-cleartext-http-url
        _ = URL(string: "http://localhost:8080/health")
        // ok: swift-cleartext-http-url
        _ = URL(string: "http://127.0.0.1:9000/debug")
    }

    // ok: swift-cleartext-http-url
    static let xmlNamespace = "http://www.w3.org/2000/svg"
}
